frontend/lib/src/components/elements/ArrowVegaLiteChart/useVegaEmbed.ts

```diff
@@ -0,0 +1,285 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { RefObject, useCallback, useEffect, useRef } from "react"
+
+import { truthy, View as VegaView } from "vega"
+import embed from "vega-embed"
+import { expressionInterpreter } from "vega-interpreter"
+
+import { useFormClearHelper } from "@streamlit/lib/src/components/widgets/Form"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+import { Quiver } from "@streamlit/lib/src/dataframes/Quiver"
+import { logMessage, logWarning } from "@streamlit/lib/src/util/log"
+import { notNullOrUndefined } from "@streamlit/lib/src/util/utils"
+
+import {
+  dataIsAnAppendOfPrev,
+  getDataArray,
+  getDataArrays,
+  getDataSets,
+  getInlineData,
+  VegaLiteChartElement,
+  WrappedNamedDataset,
+} from "./arrowUtils"
+import { useVegaLiteSelections } from "./useVegaLiteSelections"
+
+const DEFAULT_DATA_NAME = "source"
+
+interface UseVegaEmbedOutput {
+  createView: (
+    containerRef: RefObject<HTMLDivElement>,
+    spec: any
+  ) => Promise<VegaView | null>
+  updateView: (
+    data: Quiver | null,
+    datasets: WrappedNamedDataset[]
+  ) => Promise<VegaView | null>
+  finalizeView: () => void
+}
+
+export function useVegaEmbed(
+  inputElement: VegaLiteChartElement,
+  widgetMgr: WidgetStateManager,
+  fragmentId?: string
+): UseVegaEmbedOutput {
+  const vegaView = useRef<VegaView | null>(null)
+  const vegaFinalizer = useRef<(() => void) | null>(null)
+  const defaultDataName = useRef<string>(DEFAULT_DATA_NAME)
+  const dataRef = useRef<Quiver | null>(null)
+  const datasetsRef = useRef<WrappedNamedDataset[]>([])
+
+  // Setup interactivity for the chart if it supports selections
+  const { maybeConfigureSelections, onFormCleared } = useVegaLiteSelections(
+    inputElement,
+    widgetMgr,
+    fragmentId
+  )
+
+  useFormClearHelper({ widgetMgr, element: inputElement, onFormCleared })
+
+  const { id: chartId, data, datasets } = inputElement
+
+  // Initialize the data and datasets refs with the current data and datasets
+  // This is predominantly used to handle the case where we want to reference
+  // these in createView before the first render.
+  useEffect(() => {
+    if (vegaView.current === null) {
+      dataRef.current = data
+      datasetsRef.current = datasets
+    }
+  }, [data, datasets])
+
+  const finalizeView = useCallback(() => {
+    if (vegaFinalizer.current) {
+      vegaFinalizer.current()
+    }
+
+    vegaFinalizer.current = null
+    vegaView.current = null
+  }, [])
+
+  const createView = useCallback(
+    async (
+      containerRef: RefObject<HTMLDivElement>,
+      spec: any
+    ): Promise<VegaView | null> => {
+      logMessage("Creating a new Vega view.")
```

Comment:
nit: I think the messages here are more of a historical artifact .. but I believe they don't really have any actual purpose (other than debugging).

Reply:
Removed